### PR TITLE
Task6 homework

### PR DESCRIPTION
**Pull request by MarijaArnotiene87**
Pls, read the comment in Program.cs for Downcasting not working

### CODE REVIEW

File: MyShop/Shop.cs [hunk 1]
```diff
@@ -2,30 +2,58 @@
...
-        public string shopName;
-        public int openingYear;
-        public int employeeCount;
-        public string city;
+        private string shopName;
+        private int openingYear;
+        private int employeeCount;
+        private string city;
+
+        public string ShopName
```
Comment (yauhenists) [inline MyShop/Shop.cs]: Where there are no logic inside the properties you can use autoproperties and private fields are not required.
Please fix it here and check other classes

File: MainProject/UABCompany.cs [hunk 2]
```diff
@@ -0,0 +1,23 @@
+﻿namespace MainProject
+{
+    public class UABCompany : Company
+    {
+        private string ownerName;
```
Comment (yauhenists) [inline MainProject/UABCompany.cs]: Naming rules: private field in class name starts with "_" and lower case letter (e.g. _ownerName)
Please fix

File: MainProject/UABCompany.cs [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+        public override void PrintCompanyEmployeesCountByCity(string city)
+        {
+            base.PrintCompanyEmployeesCountByCity(city);
+            Console.WriteLine("Counting done");
+            Console.WriteLine();
+        }
+
+        public override string GetCompanyInfo() => base.GetCompanyInfo() + $" and owner name {ownerName}";
+
+        public new void PrintCompanyName() => Console.WriteLine($"Company hidden name is {Name}");
```
Comment (yauhenists) [inline MainProject/UABCompany.cs]: I don't get the logic to print message with "hidden". If just to show how hiding works in C# - yes, this example show it, but there is no logic

File: MyShop/Shop.cs [hunk 4]
```diff
@@ -2,30 +2,58 @@
...
+        }
+
+        public string City
+        {
+            get { return city; }
+
+            private set { city = value; }
+        }
 
         public Shop(string name, int year)
```
Comment (yauhenists) [inline MyShop/Shop.cs]: Use key word "this" for constructors to avoid duplicate code. Please check other classes

File: MainProject/Program.cs [hunk 5]
```diff
@@ -6,29 +6,45 @@ internal class Program
...
+            Company optioCompany = new Company("Optio", 1900, "LT001");
             optioCompany.PrintCompanyName();
 
-            Company dellCompany = new Company { name = "Dell", foundationYear = 2015 };
+            Company dellCompany = new Company("Dell", 2015, "LT002");
             dellCompany.PrintCompanyName();
 
             Shop seikShop = new Shop("Seik accossories", 2015, 5, "Vilnius");
-            Company seikCompany = new Company("Seik", 2014, seikShop);
+            Company seikCompany = new Company("Seik", 2014, seikShop, "LT003");
```
Comment (yauhenists) [inline MainProject/Program.cs]: Please use implicit typing where it's possible

File: MainProject/Program.cs [hunk 6]
```diff
@@ -6,29 +6,45 @@ internal class Program
...
+            Console.WriteLine(baltoUABCompany.GetCompanyInfo());
+            baltoUABCompany.PrintCompanyName();
+
+            //Upcasting
+            Company baltoCompany = baltoUABCompany;
+
+            //Downcasting
+            Company almaLiteraCompany = new Company("Alma Litera", 2020, "LT007");
+            //UABCompany almaLiteraUABCompany = (UABCompany)almaLiteraCompany; ------> sprosit v chiom problema, vikidivaet Exception
+            UABCompany? almaLiteraUABCompany = almaLiteraCompany as UABCompany;
```
Comment (yauhenists) [inline MainProject/Program.cs]: What's the point of this line of the code? Any other actions?

File: MainProject/Program.cs [hunk 7]
```diff
@@ -6,29 +6,45 @@ internal class Program
...
+
+            Shop baltoShop = new Shop("Balto parduotuve", 2014, 2, "Vilnius");
+            UABCompany baltoUABCompany = new UABCompany("Balto", 2012, "LT006", "Deividas");
+            baltoUABCompany.AddShop(baltoShop);
+            baltoUABCompany.PrintCompanyEmployeesCountByCity("Vilnius");
+            Console.WriteLine(baltoUABCompany.GetCompanyInfo());
+            baltoUABCompany.PrintCompanyName();
+
+            //Upcasting
+            Company baltoCompany = baltoUABCompany;
```
Comment (yauhenists) [inline MainProject/Program.cs]: What's the point of this line of the code? Any other actions?

File: MainProject/Program.cs [hunk 8]
```diff
@@ -6,29 +6,45 @@ internal class Program
...
+
+            //Downcasting
+            Company almaLiteraCompany = new Company("Alma Litera", 2020, "LT007");
```
Comment (yauhenists) [inline MainProject/Program.cs]: Any other actions after the object is created? (I mean calling any method)

File: MyShop/Shop.cs [hunk 9]
```diff
@@ -2,30 +2,58 @@
...
-            this.employeeCount = employeeCount;
-            this.city = city;
+            ShopName = name;
+            OpeningYear = year;
+            EmployeeCount = employeeCount;
+            City = city;
         }
 
-        public string GetShopInfo() => $"Shop {shopName} was opened in {openingYear}";
+        public string GetShopInfo() => $"Shop {ShopName} was opened in {OpeningYear}";
```
Comment (yauhenists) [inline MyShop/Shop.cs:24]: If the method is created please try to use it in your project

**Review by yauhenists — CHANGES_REQUESTED** [review]

File: MainProject/Company.cs [hunk 10]
```diff
@@ -4,44 +4,66 @@ namespace MainProject
 {
     public class Company
     {
-        public string name;
-        public int foundationYear;
-        public Shop[] shops;
+        private Shop[] shops;
+        public readonly string vatNumber;
```
Comment (yauhenists) [inline MainProject/Company.cs]: Please do not make fields as public - it should be private field or public property. Also please follow naming rules

File: MainProject/UABCompany.cs [hunk 11]
```diff
@@ -0,0 +1,21 @@
+﻿namespace MainProject
+{
+    public class UABCompany : Company
+    {
+        private string _ownerName;
+
+        public UABCompany(string name, int year, string vatNumber, string ownerName) : base(name, year, vatNumber)
+        {
+            this._ownerName = ownerName;
```
Comment (yauhenists) [inline MainProject/UABCompany.cs]: Please remove key word "this" where it's redundant

File: MainProject/Program.cs [hunk 12]
```diff
@@ -6,29 +6,38 @@ internal class Program
...
-            Company dellCompany = new Company { name = "Dell", foundationYear = 2015 };
+            var dellCompany = new Company("Dell", 2015, "LT002");
             dellCompany.PrintCompanyName();
 
-            Shop seikShop = new Shop("Seik accossories", 2015, 5, "Vilnius");
-            Company seikCompany = new Company("Seik", 2014, seikShop);
+            var seikShop = new Shop("Seik accossories", 2015, 5, "Vilnius");
+            var seikCompany = new Company("Seik", 2014, seikShop, "LT003");
             seikCompany.PrintAllShops();
+            seikShop.GetShopInfo();
```
Comment (yauhenists) [inline MainProject/Program.cs]: Method returns string and it makes no sense in this line of code - no info we get in console

File: MainProject/Company.cs [hunk 13]
```diff
@@ -4,44 +4,66 @@ namespace MainProject
...
+                shops = new[] { shop };
+            }
+            else
+            {
+                Array.Resize(ref shops, shops.Length + 1);
+                shops[shops.Length - 1] = shop;
+            }
         }
 
         public void AddShop(string shopName, int year)
```
Comment (yauhenists) [inline MainProject/Company.cs:60]: Please add for PrintAllShops method writing for which company we see the shops in console

File: MainProject/Company.cs [hunk 14]
```diff
@@ -4,44 +4,66 @@ namespace MainProject
...
+            this.vatNumber = vatNumber;
         }
 
-        public Company(string name, int year, Shop[] shops)
+        static Company()
         {
-            this.name = name;
-            foundationYear = year;
-            this.shops = shops;
+            MinimumCarCount = 2;
```
Comment (yauhenists) [inline MainProject/Company.cs:38]: I don't see any point of this static field - it's not used in any method for any conditions. But let's leave it as it is

File: MainProject/Company.cs [hunk 15]
```diff
@@ -59,19 +81,24 @@ public void PrintAllShops()
...
             {
-                if (shop.city == city)
+                if (shop.City == city)
                 {
-                    companyEmployeeCount += shop.employeeCount;
+                    companyEmployeeCount += shop.EmployeeCount;
                 }
             }
 
             Console.WriteLine($"Count of shops employees: {companyEmployeeCount}");
```
Comment (yauhenists) [inline MainProject/Company.cs]: Please add also criteria (city) in output in console

Comment (yauhenists) [reply MainProject/Company.cs]: And company as well

**Review by yauhenists — CHANGES_REQUESTED** [review]

File: MainProject/Company.cs [hunk 16]
```diff
@@ -4,73 +4,102 @@ namespace MainProject
...
 
         public void PrintAllShops()
         {
-            foreach (Shop shop in shops)
+            Console.WriteLine($"Company {Name} shops: ");
+            foreach (Shop shop in _shops)
             {
                 shop.PrintShopName();
             }
+            Console.WriteLine();
```
Comment (yauhenists) [inline MainProject/Company.cs:75]: Leave empty line after brackets ({} - blocks of code - conditions, loops). Empty line before is optional

**Review by yauhenists — CHANGES_REQUESTED** [review]

File: MyShop/Shop.cs (hunk 4 above)
Comment (MarijaArnotiene87) [reply MyShop/Shop.cs]: Fixed

Comment (MarijaArnotiene87) [reply MyShop/Shop.cs]: Fixed duplicate

File: MainProject/Company.cs (hunk 10 above)
Comment (MarijaArnotiene87) [reply MainProject/Company.cs]: Fixed

File: MainProject/UABCompany.cs (hunk 11 above)
Comment (MarijaArnotiene87) [reply MainProject/UABCompany.cs]: Fixed

File: MainProject/Program.cs (hunk 12 above)
Comment (MarijaArnotiene87) [reply MainProject/Program.cs]: Fixed

File: MainProject/Company.cs (hunk 13 above)
Comment (MarijaArnotiene87) [reply MainProject/Company.cs:60]: Fixed

File: MainProject/Company.cs (hunk 14 above)
Comment (MarijaArnotiene87) [reply MainProject/Company.cs:38]: Ok

File: MainProject/Company.cs (hunk 15 above)
Comment (MarijaArnotiene87) [reply MainProject/Company.cs]: Done

File: MainProject/Company.cs (hunk 16 above)
Comment (MarijaArnotiene87) [reply MainProject/Company.cs:75]: Done

**Review by yauhenists — APPROVED** [review]

**Pull request merged**